### PR TITLE
[Clang] Add __builtin_selectvector and use it for AVX512 intrinsics

### PR DESCRIPTION
**Pull request by philnik777**
This is also very useful for generic code. For example this would allow libc++ to vectorize `{min,max,minmax}_element` without having to use platform-specific intrinsics. I've done some testing and even at `-O0` Clang compiles the code to the expected instructions for architectures where the mask vector has the same bit count as the element vector (i.e. every SIMD ISA except AVX512 that I'm aware of) as long as the comparison operation is visible.

### CODE REVIEW

**Comment by philnik777** [comment]
I don't understand why the order of emitted instructions changes based on how exactly Clang is compiled, but other than that this should be ready. Hopefully someone spots what the problem could be.

File: clang/docs/LanguageExtensions.rst [hunk 1]
```diff
@@ -3019,6 +3019,26 @@ C-style cast applied to each element of the first argument.
 
 Query for this feature with ``__has_builtin(__builtin_convertvector)``.
 
+``__builtin_selectvector``
+--------------------------
+
+``__builtin_selectvector`` is used to express generic vector element selection.
```
Comment (RKSimon) [inline clang/docs/LanguageExtensions.rst:3025]: Extend this description to explicitly describe the input/output types and mechanism - don't just rely on the code snippet (although that's a nice accompaniment): The input must all be vectors of the same same number of elements, the 2 first operands must be the same type etc. etc. (basically everything in SemaChecking).

File: clang/lib/Headers/avx512vlintrin.h [hunk 2]
```diff
@@ -232,225 +232,225 @@ typedef char __v2qi __attribute__((__vector_size__(2)));
...
 
 static __inline__ __m128i __DEFAULT_FN_ATTRS128
 _mm_mask_add_epi64(__m128i __W, __mmask8 __U, __m128i __A, __m128i __B)
 {
-  return (__m128i)__builtin_ia32_selectq_128((__mmask8)__U,
-                                             (__v2di)_mm_add_epi64(__A, __B),
-                                             (__v2di)__W);
+  return (__m128i)__builtin_selectvector((__v2di)_mm_add_epi64(__A, __B),
+                                         (__v2di)__W,
+                                         __builtin_bit_cast(__vecmask2, __U));
```
Comment (RKSimon) [inline clang/lib/Headers/avx512vlintrin.h:317]: __mmask8 is an unsigned char - how do we safely bitcast to __vecmask2 which is bool ext vector type?

**Review by RKSimon — COMMENTED** [review]
constexpr handling?

File: clang/docs/LanguageExtensions.rst [hunk 3]
```diff
@@ -3019,6 +3019,26 @@ C-style cast applied to each element of the first argument.
...
+
+**Signature**:
+
+.. code-block:: c++
+
+  template <class T, size_t N>
+  simd_vec<T, N> __builtin_selectvector(simd_vec<T, N> lhs, simd_vec<T, N> rhs,
+                                        simd_vec<bool, N> cond)
```
Comment (phoebewang) [inline clang/docs/LanguageExtensions.rst:3033]: Maybe put `cond` the first operand to match with `select` and the old X86 builtins?

File: clang/lib/Headers/avx512fintrin.h [hunk 4]
```diff
@@ -41,6 +41,22 @@ typedef long long __m512i_u __attribute__((__vector_size__(64), __aligned__(1)))
 typedef unsigned char __mmask8;
 typedef unsigned short __mmask16;
 
+#ifdef __cplusplus
+typedef bool __vecmask2 __attribute__((__ext_vector_type__(2)));
+typedef bool __vecmask4 __attribute__((__ext_vector_type__(4)));
```
Comment (phoebewang) [inline clang/lib/Headers/avx512fintrin.h:46]: We used `__mmask8` for `v2i1` and `v4i1` too. I don't we can bitcast them directly.

File: clang/lib/Headers/avx512fintrin.h [hunk 5]
```diff
@@ -41,6 +41,22 @@ typedef long long __m512i_u __attribute__((__vector_size__(64), __aligned__(1)))
...
+typedef bool __vecmask8 __attribute__((__ext_vector_type__(8)));
+typedef bool __vecmask16 __attribute__((__ext_vector_type__(16)));
+typedef bool __vecmask32 __attribute__((__ext_vector_type__(32)));
+typedef bool __vecmask64 __attribute__((__ext_vector_type__(64)));
+#else
+typedef _Bool __vecmask2 __attribute__((__ext_vector_type__(2)));
```
Comment (phoebewang) [inline clang/lib/Headers/avx512fintrin.h:52]: I'm a bit concerning about the ABI when people abuse them in function call.

**Comment by efriedma-quic** [comment]
clang already supports `?:` with a vector condition; does this add anything new on top of that?

**Review by AaronBallman — COMMENTED** [review]
Also, please be sure to add a release note about the new functionality.

**Comment by philnik777** [comment]
> clang already supports `?:` with a vector condition; does this add anything new on top of that?

This works with bool vectors. I didn't realize you could use the ternary operator, since I only tested with them. I guess we could extend the ternary operator to accept bool vectors as well. Any thoughts?

**Comment by efriedma-quic** [comment]
You mean, if all three operands are boolean vectors?  I'm surprised that doesn't already work.

**Comment by philnik777** [comment]
> You mean, if all three operands are boolean vectors? I'm surprised that doesn't already work.

No, I mean I have a vector of bools and want to select a value based on that. e.g. `declval<simd_vector<bool, 16>>() ? declval<simd_vector<int, 16>>() : declval<simd_vector<int, 16>>()`.

**Comment by efriedma-quic** [comment]
The relevant bit of code is:

```
  // The OpenCL operator with a vector condition is sufficiently
  // different to merit its own checker.
  if ((getLangOpts().OpenCL && Cond.get()->getType()->isVectorType()) ||
      Cond.get()->getType()->isExtVectorType())
    return OpenCLCheckVectorConditional(*this, Cond, LHS, RHS, QuestionLoc);
```

Maybe makes sense to relax it.

**Pull request closed without merging**